improve docker pipe tests

`go test` may fail in the `docker` pipe if the user is not logged in with a docker account with permission into the `goreleaser` org, as reported in [here](https://gophers.slack.com/archives/C3T1JU24B/p1506799845000010).

Let's fix this!

Error looks like this:
```
command [go test -cover -covermode atomic -coverprofile /var/folders/km/kqlfxy2j64dd_9gqb373c0hm0000gn/T/gotestcover-526170983 github.com/goreleaser/goreleaser/pipeline/docker -run . -timeout=2m]: exit status 1
2017/10/01 19:57:53  info building docker image     image=goreleaser/test_run_pipe:1.0.0
2017/10/01 19:57:56  info tagging docker image      image=goreleaser/test_run_pipe:1.0.0 tag=goreleaser/test_run_pipe:latest
2017/10/01 19:57:57  info pushing docker image      image=goreleaser/test_run_pipe:1.0.0
--- FAIL: TestRunPipe (8.49s)
        Error Trace:    docker_test.go:62
	Error:  	Received unexpected error exit status 1
			failed to push docker image: 
			The push refers to a repository [docker.io/goreleaser/test_run_pipe]
			5602e019ad66: Preparing
			denied: requested access to the resource is denied
			
			github.com/goreleaser/goreleaser/pipeline/docker.dockerPush
				github.com/goreleaser/goreleaser/pipeline/docker/_test/_obj_test/docker.go:186
			github.com/goreleaser/goreleaser/pipeline/docker.publish
				github.com/goreleaser/goreleaser/pipeline/docker/_test/_obj_test/docker.go:129
			github.com/goreleaser/goreleaser/pipeline/docker.process
				github.com/goreleaser/goreleaser/pipeline/docker/_test/_obj_test/docker.go:113
			github.com/goreleaser/goreleaser/pipeline/docker.doRun
				github.com/goreleaser/goreleaser/pipeline/docker/_test/_obj_test/docker.go:66
			github.com/goreleaser/goreleaser/pipeline/docker.Pipe.Run
				github.com/goreleaser/goreleaser/pipeline/docker/_test/_obj_test/docker.go:42
			github.com/goreleaser/goreleaser/pipeline/docker.TestRunPipe
				/Users/jorin/go/src/github.com/goreleaser/goreleaser/pipeline/docker/docker_test.go:62
			testing.tRunner
				/usr/local/Cellar/go/1.9/libexec/src/testing/testing.go:746
			runtime.goexit
				/usr/local/Cellar/go/1.9/libexec/src/runtime/asm_amd64.s:2337
		
FAIL
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -6,7 +6,6 @@ services:
   - docker
 install:
   - make setup
-  - docker login -u="$DOCKER_USERNAME" -p="$DOCKER_PASSWORD"
   - gem install fpm
   - sudo apt-get update
   - sudo apt-get install --yes snapd rpm
@@ -17,6 +16,7 @@ script:
   - test -n "$TRAVIS_TAG" || go run main.go --skip-validate --skip-publish
 after_success:
   - bash <(curl -s https://codecov.io/bash)
+  - docker login -u="$DOCKER_USERNAME" -p="$DOCKER_PASSWORD"
   - test -n "$TRAVIS_TAG" && go run main.go
 notifications:
   email: false

--- a/.travis.yml
+++ b/.travis.yml
@@ -16,7 +16,7 @@ script:
   - test -n "$TRAVIS_TAG" || go run main.go --skip-validate --skip-publish
 after_success:
   - bash <(curl -s https://codecov.io/bash)
-  - docker login -u="$DOCKER_USERNAME" -p="$DOCKER_PASSWORD"
+  - test -n "$TRAVIS_TAG" && docker login -u="$DOCKER_USERNAME" -p="$DOCKER_PASSWORD"
   - test -n "$TRAVIS_TAG" && go run main.go
 notifications:
   email: false

--- a/pipeline/docker/docker_test.go
+++ b/pipeline/docker/docker_test.go
@@ -7,12 +7,31 @@ import (
 	"path/filepath"
 	"testing"
 
+	"github.com/apex/log"
+
 	"github.com/goreleaser/goreleaser/config"
 	"github.com/goreleaser/goreleaser/context"
 	"github.com/goreleaser/goreleaser/pipeline"
 
 	"github.com/stretchr/testify/assert"
 )
+
+func killAndRm() {
+	_ = exec.Command("docker", "kill", "registry").Run()
+	_ = exec.Command("docker", "rm", "registry").Run()
+}
+
+func TestMain(m *testing.M) {
+	killAndRm()
+	if err := exec.Command(
+		"docker", "run", "-d", "-p", "5000:5000", "--restart=always", "--name", "registry", "registry:2",
+	).Run(); err != nil {
+		log.WithError(err).Fatal("failed to start docker registry")
+	}
+	code := m.Run()
+	killAndRm()
+	os.Exit(code)
+}
 
 func TestRunPipe(t *testing.T) {
 	folder, err := ioutil.TempDir("", "archivetest")
@@ -24,8 +43,8 @@ func TestRunPipe(t *testing.T) {
 	_, err = os.Create(binPath)
 	assert.NoError(t, err)
 	var images = []string{
-		"goreleaser/test_run_pipe:1.0.0",
-		"goreleaser/test_run_pipe:latest",
+		"localhost:5000/goreleaser/test_run_pipe:1.0.0",
+		"localhost:5000/goreleaser/test_run_pipe:latest",
 	}
 	// this might fail as the image doesnt exist yet, so lets ignore the error
 	for _, img := range images {
@@ -39,7 +58,7 @@ func TestRunPipe(t *testing.T) {
 			Dist:        dist,
 			Dockers: []config.Docker{
 				{
-					Image:      "goreleaser/test_run_pipe",
+					Image:      "localhost:5000/goreleaser/test_run_pipe",
 					Goos:       "linux",
 					Goarch:     "amd64",
 					Dockerfile: "testdata/Dockerfile",
@@ -47,7 +66,7 @@ func TestRunPipe(t *testing.T) {
 					Latest:     true,
 				},
 				{
-					Image:      "goreleaser/test_run_pipe_nope",
+					Image:      "localhost:5000/goreleaser/test_run_pipe_nope",
 					Goos:       "linux",
 					Goarch:     "amd64",
 					Dockerfile: "testdata/Dockerfile",
@@ -70,7 +89,9 @@ func TestRunPipe(t *testing.T) {
 	// the test_run_pipe_nope image should not have been created, so deleting
 	// it should fail
 	assert.Error(t,
-		exec.Command("docker", "rmi", "goreleaser/test_run_pipe_nope:1.0.0").Run(),
+		exec.Command(
+			"docker", "rmi", "localhost:5000/goreleaser/test_run_pipe_nope:1.0.0",
+		).Run(),
 	)
 }
 

--- a/pipeline/docker/docker_test.go
+++ b/pipeline/docker/docker_test.go
@@ -17,6 +17,7 @@ import (
 )
 
 func killAndRm() {
+	log.Info("killing registry")
 	_ = exec.Command("docker", "kill", "registry").Run()
 	_ = exec.Command("docker", "rm", "registry").Run()
 }
@@ -24,13 +25,12 @@ func killAndRm() {
 func TestMain(m *testing.M) {
 	killAndRm()
 	if err := exec.Command(
-		"docker", "run", "-d", "-p", "5000:5000", "--restart=always", "--name", "registry", "registry:2",
+		"docker", "run", "-d", "-p", "5000:5000", "--name", "registry", "registry:2",
 	).Run(); err != nil {
 		log.WithError(err).Fatal("failed to start docker registry")
 	}
-	code := m.Run()
-	killAndRm()
-	os.Exit(code)
+	defer killAndRm()
+	os.Exit(m.Run())
 }
 
 func TestRunPipe(t *testing.T) {

